test/api_integration/apis/telemetry/utils/schema_to_config_schema.ts

```diff
@@ -0,0 +1,138 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema, ObjectType, Type } from '@kbn/config-schema';
+import { get } from 'lodash';
+import { set } from '@elastic/safer-lodash-set';
+import type { AllowedSchemaTypes } from 'src/plugins/usage_collection/server';
+
+/**
+ * Type that defines all the possible values that the Telemetry Schema accepts.
+ * These types definitions are helping to identify earlier the possible missing `properties` nesting when
+ * manually defining the schemas.
+ */
+export type TelemetrySchemaValue =
+  | {
+      type: AllowedSchemaTypes | string;
+    }
+  | { type: 'array'; items: TelemetrySchemaValue }
+  | TelemetrySchemaObject;
+
+export interface TelemetrySchemaObject {
+  properties: Record<string, TelemetrySchemaValue>;
+}
+
+function isOneOfCandidate(
+  schemas: Array<Type<unknown>>
+): schemas is [Type<unknown> | Type<unknown>] {
+  return schemas.length === 2;
+}
+
+/**
+ * Converts each telemetry schema value to the @kbn/config-schema equivalent
+ * @param value
+ */
+function valueSchemaToConfigSchema(value: TelemetrySchemaValue): Type<unknown> {
+  if ('properties' in value) {
+    const { DYNAMIC_KEY, ...properties } = value.properties;
+    const schemas: Array<Type<unknown>> = [objectSchemaToConfigSchema({ properties })];
+    if (DYNAMIC_KEY) {
+      schemas.push(schema.recordOf(schema.string(), valueSchemaToConfigSchema(DYNAMIC_KEY)));
+    }
+    return isOneOfCandidate(schemas) ? schema.oneOf(schemas) : schemas[0];
+  } else {
+    const valueType = value.type; // Copied in here because of TS reasons, it's not available in the `default` case
+    switch (value.type) {
+      case 'boolean':
+        return schema.boolean();
+      case 'keyword':
+      case 'text':
+      case 'date':
+        return schema.string();
+      case 'byte':
+      case 'double':
+      case 'float':
+      case 'integer':
+      case 'long':
+      case 'short':
+        // Some plugins return `null` when there is no number to report
+        return schema.oneOf([schema.number(), schema.literal(null)]);
+      case 'array':
+        if ('items' in value) {
+          return schema.arrayOf(valueSchemaToConfigSchema(value.items));
+        }
+      default:
+        throw new Error(
+          `Unsupported schema type ${valueType}. Did you forget to wrap your object definition in a nested 'properties' field?`
+        );
+    }
+  }
+}
+
+function objectSchemaToConfigSchema(objectSchema: TelemetrySchemaObject): ObjectType {
+  return schema.object(
+    Object.fromEntries(
+      Object.entries(objectSchema.properties).map(([key, value]) => {
+        try {
+          return [key, schema.maybe(valueSchemaToConfigSchema(value))];
+        } catch (err) {
+          err.failedKey = [key, ...(err.failedKey || [])];
+          throw err;
+        }
+      })
```

Comment:
Should we add tests on this test util 😅 ?

Reply:
Who tests the tester of the tester? 😅 
I think it's worth adding them. I'll add some unit tests to this tool. Thanks for bringing it up!

Reply:
Done! I actually found an edge case we didn't cover in `flatKeys`